<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобрадотка-данных" data-toc-modified-id="Предобрадотка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобрадотка данных</a></span><ul class="toc-item"><li><span><a href="#Пропуски-и-артефакты" data-toc-modified-id="Пропуски-и-артефакты-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Пропуски и артефакты</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Дубликаты" data-toc-modified-id="Дубликаты-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Дубликаты</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Проверка-гипотез" data-toc-modified-id="Проверка-гипотез-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Проверка гипотез</a></span><ul class="toc-item"><li><span><a href="#Влияние-семейного-положения" data-toc-modified-id="Влияние-семейного-положения-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Влияние семейного положения</a></span></li><li><span><a href="#Влияние-количества-детей" data-toc-modified-id="Влияние-количества-детей-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Влияние количества детей</a></span></li><li><span><a href="#Дополнительные-выводы" data-toc-modified-id="Дополнительные-выводы-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Дополнительные выводы</a></span></li></ul></li><li><span><a href="#Результаты-исследования" data-toc-modified-id="Результаты-исследования-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Результаты исследования</a></span></li></ul></div>

 # Исследование платёжеспособности заёмщиков

Кредитный отдел банка поручил проанализировать влияние некоторы факторов на платёжеспособность заёмщиков. Отдел предоставил данные о платёжеспособности клиентов.

Описание данных:
* `children` — количество детей в семье
* `days_employed` — общий трудовой стаж в днях
* `dob_years` — возраст клиента в годах
* `education` — уровень образования клиента
* `education_id` — идентификатор уровня образования
* `family_status` — семейное положение
* `family_status_id` — идентификатор семейного положения
* `gender` — пол клиента
* `income_type` — тип занятости
* `debt` — имел ли задолженность по возврату кредитов
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита

**Цель исследования** - проверить две гипотезы:

1. Семейное положение клиента влияет на факт своевременного погащения кредита.
2. Количество детей у заёмщика влияент на факт возврата кредита в срок.

**Ход исследования:**

1. Обзор данных
2. Предобработка данных
3. Проверка гипотез

## Обзор данных

Для ознакомления с таблицей импортируем библиотеку `pandas`  и прочитаем предоставленный файл. Выведем на экран первые 10 строк, чтобы составить первое впечатление о полученных данных:

In [1]:
import pandas as pd

data = pd.read_csv('//censored//')
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Столбцы названы корректно: они отражают суть представленных в них данных и записаны в "змеином регистре". Однако в самих данных уже сейчас видны неверные значения (отрицательное число дней трудового стажа в стоблце `days_employed`) и неявные дубликаты (в стоблцах `education` и `purpose`).
Кроме того, таблица слишком "громоздкая" для анализа. Удобнее будет разделить её на основную таблицу и несколько словарей.

Посмотрим общую информацию о таблице:

In [2]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Типы данных в стоблцах в целом выбраны корректно, однако можно заменить вещественный тип в столбцах `days_employed` и `total_income` на целочисленный.
Всего в таблице 21525 строк. Но в данных о трудовом стаже `days_employed` и ежемесячном доходе `total_income` есть пропуски, причём одинаковое количество.

Вычислим, какую долю составляют пропуски от общего числа значений в каждом из данных стобцов:

In [3]:
display(pd.DataFrame(round((data.isna().mean()*100),2), columns=['NaNs, %']).style.format(
    '{:.2f}').background_gradient('coolwarm'))

Таким образом, пропущенные значения составляют чуть больше 10% от общего числа значений. в каждом из двух столбцов. Учитывая, что количество пропусков в данных столбцах одинаково, стоит проверить, встречаются ли пропуски в одних и тех же строках:

In [4]:
len(data.loc[(data['days_employed'].isna()) & (data['total_income'].isna())]), data['days_employed'].isna().sum()

(2174, 2174)

Получется, что все пропущенные значения в столбцах `days_emplyed` и `total_income` встречаются в одних и тех же строках.

Действительно, пропуски в столбцах `days_employed` и `total_income` встречаются в одних и тех же строках. В связи с этим их появление может быть обусловлено следующими причинами:
* Столбцы `days_employed` и `total_income` появились в таблице через какое-то время после её создания, в связи с этим данные характеристики отстутсвуют у клиентов, данные о которых были занесены в таблицу до появления этих столбцов.
* Определённый/ые работник/и по каким-то причинам пропустили две эти характеристики и не внесли их в базу.
* Произошёл технический сбой, и данные были утеряны.

**Выводы:**

1. Перед тем как проверять гипотезы необходимо избавиться от артефактов в стоблце `days_employed` и проверить другие столбцы на артефакты.
2. Также надо заполнить пустые значения в столбцах `days_employed` и `total_income`.
3. Затем удалим явные и неявные дубликаты.
4. Изменим тип в столбцах `days_employed` и `total_income` с float на int.
5. Создадим несколько более ёмких датафреймов на основе исходного.

## Предобрадотка данных

### Пропуски и артефакты

Хотя столбы `days_employed` и `total_income` не имеют непосредственного значения для тестирования поставленных гипотез, всё же приведём данные в этих столбцах в порядок.

Для начала разберёмся со столбцом `total_income`. Поскольку столбец содержит количественные значения, заполнять пропуски в нём следует медианным или средним значением.

Посмотрим расхождение между этими двумя значениями:

In [5]:
avg_income = data['total_income'].mean()
median_income = data['total_income'].median()

round(avg_income), round(median_income)

(167422, 145018)

Так, среднее значение превышает медианное более чем на 20000. Проверим, есть ли в таблице "выдающиеся значения" дохода:

In [6]:
data.sort_values(by='total_income', ascending=False)['total_income'].round().head(15)

12412    2265604.0
19606    2200852.0
9169     1726276.0
20809    1715018.0
17178    1711309.0
17503    1597613.0
18368    1551153.0
18353    1427934.0
15268    1350246.0
11071    1286281.0
2224     1278623.0
7447     1240165.0
19813    1223042.0
15211    1172460.0
13090    1128836.0
Name: total_income, dtype: float64

У двух клиентов с самым высоким заработком месячный доход существенно выше даже, чем у их ближайших последователей, что говорит о значительном разбросе значений. В целом, у Топ-15 клиентов по уровню заработной платы месячный доход в 10 раз превышает среднее и медианное значения.

В данном случае более корректно заполнить пропуски медианным значением месячного дохода, поскольку оно лучше характеризует всю выборку, отбрасывая экстремально большие и маленькие значения.

Заполним пропуски в столбце `total_income` медианным значением с учётом категории `income_type`:

In [7]:
income_types = data['income_type'].unique()

for value in income_types:
    data.loc[data['income_type'] == value, 'total_income'] = data.loc[data['income_type'] == value, 'total_income'].fillna(
        data.loc[data['income_type'] == value, 'total_income'].median())   

#Проверим:
data['total_income'].isna().sum()

0

Отлично, теперь в столце `total_income` нет пропусков.

Также необходимо исправить значения в столбце `days_employed`. Посмотрим, сколько значений не соотверствуют действительности. Для этого выделим два критерия:
* Количество дней стажа должно быть неотрицательным числом.
* Даже если человек работал 65 лет по 360 дней, его трудовой стаж будет равен 23400 дней. Следовательно, будем отдельно анализировать все значения, превышающие данный порог.

Составим список с "подозрительными" значениями столбца `days_employed`:

In [8]:
suspicious_values = []

#Напишем функцию для заполнения данного списка:
def find_suspicious_values(value):
    if value < 0 or value > 23400:
        suspicious_values.append(value)

data['days_employed'].apply(find_suspicious_values)     
suspicious_values

[-8437.673027760233,
 -4024.803753850451,
 -5623.422610230956,
 -4124.747206540018,
 340266.07204682194,
 -926.1858308789184,
 -2879.202052139952,
 -152.7795691752705,
 -6929.865298973741,
 -2188.7564450779378,
 -4171.483646903305,
 -792.7018870609315,
 -1846.6419410560736,
 -1844.9561821875543,
 -972.3644189900732,
 -1719.9342261530542,
 -2369.999720209008,
 400281.1369125588,
 -10038.818548915877,
 -1311.6041664289485,
 -253.68516578402495,
 -1766.6441376815046,
 -272.9813850334044,
 338551.95291136915,
 363548.4893483957,
 -529.1916346594236,
 -717.2743239418727,
 335581.6685149753,
 -1682.08343808475,
 -4649.910831728614,
 -1548.63754364046,
 -4488.067031277062,
 394021.0721837952,
 -176.21668833712155,
 -6448.810859676204,
 -597.8818273051091,
 -650.5877964026586,
 -1030.3626056014652,
 -1257.4961901986044,
 -4375.681384360526,
 -1362.0417275002387,
 -1039.4517405766287,
 -2262.712303593806,
 -2689.1372742032745,
 -3341.0678863600638,
 -1181.443228272809,
 353731.4323382981,
 -695

Как видно, в столбце очень много артефактов. Посчитаем количество каждого типа с помощью цикла:

In [9]:
count_negative_values = 0 #количество отрицательных значений в столбце
count_too_large_values = 0 #количество слишком больших значений в столбце

for value in suspicious_values:
    if value < 0:
        count_negative_values += 1
    else:
        count_too_large_values += 1

count_negative_values, count_too_large_values

(15906, 3445)

Таким образом, в столбце есть 15906 отрицательных значений и 3445 положительных значений, которые всё равно не соответствуют действительности (стаж в 900 лет без выходных требует невероятного трудолюбия). Посчитаем долю каждого из типов артефактов во всех значениях столбца (19351):

In [10]:
round(count_negative_values/19351, 3), round(count_too_large_values/19351, 3)

(0.822, 0.178)

Выходит, что столбец `days_employed` целиком состоит из артефактов!
В случае с отрицательными значениями, причиной их возникновения могли послужить технические ошибки. В связи с этим можно допустить, что значения, написанные со знаком минус, должны быть положительными. Запишем их по модулю:

In [11]:
data['days_employed'] = data['days_employed'].abs()

#Так как отрицательные значения трудового стажа фигурировали в первых 10 строках таблицы, проверим их ещё раз:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   СРЕДНЕЕ             1   
8         2    6929.865299         35    ВЫСШЕЕ             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Отрицательные значения ликвидированы.

Теперь поподробнее изучим, у каких категорий людей встречаются экстремально большие значения трудового стажа:

In [12]:
data.loc[data['days_employed'] > 23400, ['days_employed', 'dob_years', 'gender', 'income_type']].head(15)

days_employed  dob_years gender income_type
4   340266.072047         53      F   пенсионер
18  400281.136913         53      F   пенсионер
24  338551.952911         57      F   пенсионер
25  363548.489348         67      M   пенсионер
30  335581.668515         62      F   пенсионер
35  394021.072184         68      M   пенсионер
50  353731.432338         63      F   пенсионер
56  370145.087237         64      F   пенсионер
71  338113.529892         62      F   пенсионер
78  359722.945074         61      M   пенсионер
86  383933.549664         64      M   пенсионер
87  337659.535004         53      F   пенсионер
88  373129.789654         63      F   пенсионер
98  364906.205736         54      F   пенсионер
99  346541.618895          0      F   пенсионер

Так, мы выяснили, что такой огромный стаж работы, встречается в таблице у пенсионеров. Заменим экстремальные значения трудового стажа на максимально возможные реалистичные. Будем считать, что клиенты с запредельно большим трудовым стажем начали работать с 18 лет и работали 360 дней в году.

Напишем функцию, которая заменит артефакты на соответствующие значения. Но сначала проверим корректность значений столбца `dob_years`:

In [13]:
min(data['dob_years']), max(data['dob_years'])

(0, 75)

Очевидно, что заёщику не может быть ноль лет, поэтому проверим, сколько в таблице заёмщиков моложе 18:

In [14]:
data[data['dob_years'] < 18]['dob_years'].count()

101

Так, у многих клиентов в таблице возраст явно не соответствует действительности.

Проверим, информация о каких категориях клиентов есть в датафрейме:

In [15]:
data['income_type'].sort_values().unique()

array(['безработный', 'в декрете', 'госслужащий', 'компаньон',
       'пенсионер', 'предприниматель', 'сотрудник', 'студент'],
      dtype=object)

Найдём средний возраст заёмщиков по двум категориям - люди трудоспособного возраста и пенсионеры.

In [16]:
display(round(data.loc[data['income_type'] == 'пенсионер', 'dob_years'].mean()))
round(data.loc[data['income_type'] != 'пенсионер', 'dob_years'].mean())

59

40

Заполним "странные" значения возраста средними для каждой категории заёмщиков:

In [17]:
data.loc[(data['dob_years'] < 18) & (data['income_type'] == 'пенсионер'), 'dob_years'] = 59
data.loc[(data['dob_years'] < 18) & (data['income_type'] != 'пенсионер'), 'dob_years'] = 40

#Проверим:
data[data['dob_years'] == 0]['dob_years'].count()

0

Неправдоподобных значений в колонке возраст больше нет, значит, можно писать функцию для замены артефактов в столбце `days_employed`.

In [18]:
def lower_days_employed(row):
    
    if row['days_employed'] is not None:
        if row['days_employed'] > 23400:
            row['days_employed'] = (row['dob_years']-18)*360
            return row['days_employed']
        else:
            return row['days_employed']
    else:
        return row['days_employed']

#Проверим функцию на строке:
row = [1234567890, 30]
columns = ['days_employed', 'dob_years']
row = pd.Series(data=row, index=columns)
lower_days_employed(row)

display(row)

#И на датафрейме:
rows = {'days_employed':[1234567890, 5678],
    'dob_years':[67, 40]}
rows = pd.DataFrame(data=rows)
rows['days_employed'] = rows.apply(lower_days_employed, axis=1)

rows

days_employed    4320
dob_years          30
dtype: int64

days_employed  dob_years
0          17640         67
1           5678         40

In [19]:
#Теперь можно применять функцию к датафрейму:
data['days_employed'] = data.apply(lower_days_employed, axis=1)

data.head(10)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0   12600.000000         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   СРЕДНЕЕ             1   
8         2    6929.865299         35    ВЫСШЕЕ             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Данные выглядят корректно. Но на всякий случай проверим, осталились ли в столбце нереалистичные значения:

In [20]:
data.loc[(data['days_employed']<0)|(data['days_employed']>23400)]['days_employed'].count()

0

Артефактов в столбце больше нет!

Теперь заполним пропуски. Для этого рассчитаем среднее и медианное значения трудового стажа:

In [21]:
avg_days_employed = data['days_employed'].mean()
median_days_employed = data['days_employed'].median()

round(avg_days_employed), round(median_days_employed)

(4588, 2194)

Среднее значение трудового стажа более, чем в 2 раза превышает медианное. Учитывая значительный разброс данных, лучше опять же заполнить пропуски медианными значениями. Но заполнять пропуски в столбце будем с учётом возраста клиента, поскольку от него сильно зависит стаж:

In [22]:
data['age_category'] = pd.cut(data['dob_years'], [18, 30, 40, 50, 60, 75])
data['days_employed'] = data['days_employed'].fillna(data.groupby('age_category')['days_employed'].transform('median'))
data = data.drop('age_category', axis=1)

#Проверим:
data['days_employed'].isna().sum()

0

Все пропуски в таблице искоренены!

Посмотрим, какие значения принимает столбец `children`:

In [23]:
data['children'].sort_values().unique()

array([-1,  0,  1,  2,  3,  4,  5, 20])

Оказалось, что в данном столбце тоже есть как минимум один артефакт - -1. 20 детей у человека быть, конечно, может, но это единичный случай. Проверим количество таких людей в таблице. 

In [24]:
data[data['children'] == 20]['children'].count()

76

Такая высокая рождаемость уже очень маловероятна. Значит, 20 - это тоже артефакт. Посчитаем количество строк с артифактами в столбце `children`.

In [25]:
data.loc[(data['children'] == -1) | (data['children'] == 5)]['children'].count()

56

Поскольку дополнительных данных о количестве детей у клиентов нет, а число аномальных строк не так велико, чтобы повлиять на исход исследования, удалим их:

In [26]:
data = data.drop(data[(data['children'] == -1) | (data['children'] == 20)].index)

#Проверим:
data['children'].sort_values().unique()

array([0, 1, 2, 3, 4, 5])

Отлично, артфекты полностью удалены!

### Изменение типов данных

Сделаем количественные данные более читабельными, заменив тип данных в столбцах `days_employed` и `total_income` с вещественного на целочисленный.

In [27]:
data['days_employed'] = data['days_employed'].astype('int')
data['total_income'] = data['total_income'].astype('int')

data.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   Среднее             1   
3         3           4124         32   среднее             1   
4         0          12600         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Теперь данные выглядят понятнее.

### Дубликаты

Для начала найдём количество полностью повторяющихся строк таблицы.

In [28]:
data.duplicated().sum()

54

Удалим повторяющиеся строки:

In [29]:
data = data.drop_duplicates().reset_index(drop=True)

Теперь займёмся поиском неявных дубликатов. Они могут возникнуть в столбцах `education`, `family_status`, `gender` и `purpose`.

Начнём рассмотрение со столба `education`.

In [30]:
data['education'].sort_values().unique()

array(['ВЫСШЕЕ', 'Высшее', 'НАЧАЛЬНОЕ', 'НЕОКОНЧЕННОЕ ВЫСШЕЕ',
       'Начальное', 'Неоконченное высшее', 'СРЕДНЕЕ', 'Среднее',
       'УЧЕНАЯ СТЕПЕНЬ', 'Ученая степень', 'высшее', 'начальное',
       'неоконченное высшее', 'среднее', 'ученая степень'], dtype=object)

Так, одни и те же стадии образования записаны в разных регистрах, что вызывает появление неявных дубликатов. Причиной этого может быть как человеческий фактор, так и техническая ошибка. Приведём все значения к нижнему регистру:

In [31]:
data['education'] = data['education'].str.lower()

#Пересчитаем количество явных дубликатов:
data.duplicated().sum()

17

In [32]:
#Избавимся от новых дубликатов:
data = data.drop_duplicates().reset_index(drop=True)

#Проверим уникальные значения столбца education:
data['education'].sort_values().unique()

array(['высшее', 'начальное', 'неоконченное высшее', 'среднее',
       'ученая степень'], dtype=object)

Так, в столбце `education` неявных дубликатов бельше нет.

Проверим столбец `family_status`:

In [33]:
data['family_status'].sort_values().unique()

array(['Не женат / не замужем', 'в разводе', 'вдовец / вдова',
       'гражданский брак', 'женат / замужем'], dtype=object)

В данном столбце все значения уникальные, неявных дубликатов нет, однако можно привести все записи к нижнему регистру.

In [34]:
data['family_status'] = data['family_status'].str.lower()

data['family_status'].sort_values().unique()

array(['в разводе', 'вдовец / вдова', 'гражданский брак',
       'женат / замужем', 'не женат / не замужем'], dtype=object)

Проверим на предмет неявных дубликатов столбец `gender`:

In [35]:
data['gender'].sort_values().unique()

array(['F', 'M', 'XNA'], dtype=object)

В столбце есть три неповторяющиеся значения - женский пол, мужской пол, пол не указан. Значит, дубликатов нет.

Посчитаем количество клиентов, пол которых не указан:

In [36]:
data[data['gender'] == 'XNA']['gender'].count()

1

Такой человек всего один, значит, данная аномалия не может повлиять на результаты анализа данных. Тем более, возможно,  банком предполагалось, что человек может не указывать свой пол, поэтому можно оставить.

Осталось проверить на предмет неявных дубликатов столбец `purpose`:

In [37]:
data['purpose'].sort_values().unique()

array(['автомобили', 'автомобиль', 'высшее образование',
       'дополнительное образование', 'жилье',
       'заняться высшим образованием', 'заняться образованием',
       'на покупку автомобиля', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля', 'на проведение свадьбы',
       'недвижимость', 'образование', 'операции с жильем',
       'операции с коммерческой недвижимостью',
       'операции с недвижимостью', 'операции со своей недвижимостью',
       'покупка жилой недвижимости', 'покупка жилья',
       'покупка жилья для сдачи', 'покупка жилья для семьи',
       'покупка коммерческой недвижимости', 'покупка недвижимости',
       'покупка своего жилья', 'получение высшего образования',
       'получение дополнительного образования', 'получение образования',
       'приобретение автомобиля', 'профильное образование',
       'ремонт жилью', 'свадьба', 'свой автомобиль',
       'сделка с автомобилем', 'сделка с подержанным автомобилем',
       'строительство 

В данном столбце присутствуют схожие цели, разные по степени конкретизации, которые можно подразделить на 4 большие группы - операции с автомобилей, операции с недвижимостью, проведение свадьбы и получение образования. Поскольку более конкретные данные могут быть нужны для других аналитических целей, разумно будет не заменять значения в столбце `purpose`, а создать столбец с категориями целей.

### Категоризация данных

Для начала сделаем отдельные "словари" для характеристики образования и семейного статуса, чтобы данные было легче читать:

In [38]:
#Создадим словарь education_dict:
education_dict = data[['education_id', 'education']]
education_dict = education_dict.drop_duplicates().reset_index(drop=True)

education_dict = education_dict.sort_values(by='education_id')
education_dict

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

In [39]:
#Создадим словарь family_dict:
family_dict = data[['family_status_id', 'family_status']]
family_dict = family_dict.drop_duplicates().reset_index(drop=True)

family_dict = family_dict.sort_values(by='family_status_id')
family_dict

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  не женат / не замужем

Удалим столбцы `education` и `family_status` из изначального датафрейма, чтобы он выглядел менее громоздким:

In [40]:
data = data.drop(['education', 'family_status'], axis=1)

data.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0          12600         53             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

Теперь, когда таблица стала посвободнее, с помощью функции создадим новый столбец `purpose_category` для обобщения целей получения кредита в несколько вышеупомянутых категорий.

In [41]:
def group_purposes(purpose):
    
    if purpose.find('автомобил') != -1:
        return 'операции с автомобилем'
    elif purpose.find('недвижимост') != -1 or purpose.find('жиль') != -1:
        return 'операции с недвижимостью'
    elif purpose.find('свадьб') != -1:
        return 'проведение свадьбы'
    elif purpose.find('образовани') != -1:
        return 'получение образования'

#Проверим функцию:
group_purposes('ремонт автомобиля')

'операции с автомобилем'

In [42]:
#Применим функцию к датафрейму:
data['purpose_category'] = data['purpose'].apply(group_purposes)

#Посмотрим на результат:
data.head(10)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0          12600         53             1                 1      F   
5         0            926         27             0                 1      M   
6         0           2879         43             0                 0      F   
7         0            152         50             1                 0      M   
8         2           6929         35             0                 1      F   
9         0           2188         41             1                 0      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   
5   компаньон     0        255763               покупка жилья   
6   компаньон     0        240525           операции с жильем   
7   сотрудник     0        135823                 образование   
8   сотрудник     0         95856       на проведение свадьбы   
9   сотрудник     0        144425     покупка жилья для семьи   

           purpose_category  
0  операции с недвижимостью  
1    операции с автомобилем  
2  операции с недвижимостью  
3     получение образования  
4        проведение свадьбы  
5  операции с недвижимостью  
6  операции с недвижимостью  
7     получение образования  
8        проведение свадьбы  
9  операции с недвижимостью

Для дальнейшего анализа также сгруппируем клиентов в несколько категорий по уровню дохода. Для этого напишем функцию:

In [43]:
def group_by_income(income):
    
    if income <= 30000:
        return 'E'
    elif 30000 < income <= 50000:
        return 'D'
    elif 50000 < income <= 200000:
        return 'C'
    elif 200000 < income <= 1000000:
        return 'B'
    else:
        return 'A'

#Проверка:
group_by_income(150000)

'C'

In [44]:
#Создадим новый столбец с категориями дохода клиентов:
income_category = [] #список для категорий дахода

income_category = data['total_income'].apply(group_by_income) #заполним список значениями
data.insert(9, 'total_income_category', income_category) #вставим столбец с категориями дохода клиентов рядом со столбцом total_income
data.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0          12600         53             1                 1      F   

  income_type  debt  total_income total_income_category  \
0   сотрудник     0        253875                     B   
1   сотрудник     0        112080                     C   
2   сотрудник     0        145885                     C   
3   сотрудник     0        267628                     B   
4   пенсионер     0        158616                     C   

                      purpose          purpose_category  
0               покупка жилья  операции с недвижимостью  
1     приобретение автомобиля    операции с автомобилем  
2               покупка жилья  операции с недвижимостью  
3  дополнительное образование     получение образования  
4             сыграть свадьбу        проведение свадьбы

Теперь ещё раз посмотрим информацию о таблице:

In [45]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21331 entries, 0 to 21330
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   children               21331 non-null  int64 
 1   days_employed          21331 non-null  int64 
 2   dob_years              21331 non-null  int64 
 3   education_id           21331 non-null  int64 
 4   family_status_id       21331 non-null  int64 
 5   gender                 21331 non-null  object
 6   income_type            21331 non-null  object
 7   debt                   21331 non-null  int64 
 8   total_income           21331 non-null  int64 
 9   total_income_category  21331 non-null  object
 10  purpose                21331 non-null  object
 11  purpose_category       21331 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


Таблица готова к анализу.

## Проверка гипотез

### Влияние семейного положения

Согласно первой гипотезе, семейный статус влияет на возврат кредита в срок.
Наличие задолженности по кредиту показывает булевая переменная `debt` (1 - если была задолженность).

Посчитаем общее количество клиентов, не погасивших кредит своевременно:

In [46]:
data['debt'].sum(), data['debt'].mean()

(1732, 0.08119638085415593)

Количество клиентов, имевших задолженность по возврату кредитов, составляет 1741 человек, или 8,12% от общего числа заёмщиков.

Для проверки данной гипотезы сгруппируем данные по семейному статусу, чтобы посмотреть, сколько клиентов из разных групп по семейному статусу имели задолженность по кредиту.

In [47]:
def make_pivot(category):
    pivot = data.pivot_table(index=category, values='debt', aggfunc=['sum', 'count', 'mean'])
    pivot = pivot.round(4)
    pivot = pivot.rename(
    columns={'sum':'debtors', 'count':'total', 'mean':'percentage'})
    pivot.columns = pivot.columns.droplevel([1])

    return pivot

fam_status_table = make_pivot('family_status_id')
fam_status_table

debtors  total  percentage
family_status_id                            
0                     927  12261      0.0756
1                     385   4134      0.0931
2                      63    951      0.0662
3                      84   1189      0.0706
4                     273   2796      0.0976

In [48]:
#Объединим получившуюся таблицу со словарём family_dict:
fam_status_dict = family_dict.merge(fam_status_table, on='family_status_id', how='left')

fam_status_dict

family_status_id          family_status  debtors  total  percentage
0                 0        женат / замужем      927  12261      0.0756
1                 1       гражданский брак      385   4134      0.0931
2                 2         вдовец / вдова       63    951      0.0662
3                 3              в разводе       84   1189      0.0706
4                 4  не женат / не замужем      273   2796      0.0976

Получается, что больше всего клиентов с задолженностью по кредиту средиты женатых/замужних людей, а меньше всего - среди вдов/цов и тех, кто в разводе. Тем не менее это также самая многочисленная и самые малочисленные группы клиентов соответственно.

Однако в относительном выражении самая большая доля людей с задолженностью по возврату кредита в группе тех, кто не женат/не замужем (около 10%), а среди женатых/замужних доля должников такая же, как и по всему датафрейму (около 7,5%). Второе место по количеству должников занимает группа клиентов, находящихся в гражданском браке (9,3%). При этом меньше всего должников среди разведённых и вдов/цов (по 6,6-7%).

В данном случае клиентов из всех групп по семейному статусу можно объединить в две агрегированные группы - тех, у кого есть партнёр (первые две группы по семейному статусу), и тех, у кого нет партнёра (три оставшиеся группы). Поскольку, неочевидно, как объяснить различия в относительном числе должников внутри каждой из этих агрегированных групп, можно предположить, что причинно-следственной связи между семейным статусом и своевременным возратом кредита провести нельзя. Здесь могут играть роль скорее косвенные факторы (получение вдовой/цом наследства, разведёнными супругами - алиментов или части имущества и т.д.).

Учитывая это, можно сказать, что семейное положение не оказывает прямого влияния на факт своевременного погашения кредита, т. е. первая гипотеза подтверждена лишь частично.

### Влияние количества детей

Теперь рассмотрим, влияет ли количество детей на факт погашения кредита в срок. Для этого сделаем аналогичную таблицу той, что была в предыдущем пункте:

In [49]:
make_pivot('children')

debtors  total  percentage
children                            
0            1063  14091      0.0754
1             444   4808      0.0923
2             194   2052      0.0945
3              27    330      0.0818
4               4     41      0.0976
5               0      9      0.0000

В данном случае меньше всего должников в категории клиентов с 5 детьми (их вообще нет), однако это можно объяснить малочисленностью группы. В целом можно проследить повышательную тенденцию количества должников в группе при увеличении детей (исключением является только категория клиентов с 3 детьми, выбивающаяся из общего тренда). В связи с этим можно сделать вывод, что клиенты без детей с большей вероятностью вернут кредит в срок (задолженность из них имели лишь 7,5%).

### Дополнительные выводы

Посмотрим, есть ли зависимость между фактом своевременного возврата кредита и ещё двумя факторами: 1) уровнем дохода и 2) целью кредита.

Сначала проверим связь с уровнем дохода. Для этого построим сводную таблицу количества должников в группах по уровню дохода.

In [50]:
make_pivot('total_income_category')

debtors  total  percentage
total_income_category                            
A                            2     25      0.0800
B                          354   5014      0.0706
C                         1353  15921      0.0850
D                           21    349      0.0602
E                            2     22      0.0909

Таким, образом никакого тренда относительно вероятность невозврата кредита в срок в зависимости от дохода клиента не прослеживается. Как ни странно, меньше всего процент должников в группе клиентов с доходом от 30 до 50 тысяч рублей, т. е. во второй группе с конца (D).

Проверим, сильно ли различается средний доход среди тех, кто вовремя и не вовремя вернул кредит, внутри групп по уровню дохода (здесь среднее значение берём именно для того, чтобы учесть и экстремальные значения).

In [51]:
income_pivot = data.pivot_table(
    index=['total_income_category'], 
    columns='debt',
    values='total_income'
)
income_pivot = income_pivot.round()
income_pivot

debt                           0          1
total_income_category                      
A                      1315052.0  1615876.0
B                       286909.0   283136.0
C                       128087.0   129124.0
D                        42925.0    42851.0
E                        25883.0    24910.0

Выходит, что существенных различий в размере заработка между теми, кто своевременно и несвоевременно выплачивает кредит, внутри каждой группы по уровню дохода также нет. А в некоторых случаях (как в группах A и C) ежемесячный доход клиентов с задолженностью по кредиту даже немного выше чем, у тех, у кого задолженности нет.

Проверим, влияет ли цель кредита на его своевременное погашение.

Снова составим сводную таблицу:

In [52]:
make_pivot('purpose_category')

debtors  total  percentage
purpose_category                                    
операции с автомобилем        400   4279      0.0935
операции с недвижимостью      780  10751      0.0726
получение образования         369   3988      0.0925
проведение свадьбы            183   2313      0.0791

Таким образом, клиенты, которые берут кредит на автомобиль или образование, чаще не возвращают его в срок.

## Результаты исследования

Проверка гипотез показала, что:

1. Семейный статус не оказывает прямого влияния на факт своевременного погашения кредита, однако меньше всего должников среди вдов/цов и разведённых людей.
2. С увеличением количества детей риск не возврата кредита в срок возрастает.
3. Уровень дохода не являются значимыми факторами при определении платёжеспособности клиента.
4. Должниками чаще оказываются клиенты, взявшие кредит на автомобиль или образование.

Из рассмотренных факторов можно составить портет "идеального" заёмщика: им является вдовец/вдова или разведённый человек, у которого/ой нет детей и которому/ой кредит нужен для проведения свадьбы или операций с недвижимостью.

Таким образом, не все выдвинутые гипотезы подтвердились. Чтобы более точно оценить, от каких факторов зависит факт возвращения кредита в срок, необходимо дополнительное исследование (например, можно посмотреть условия кредитов, процентные ставки, график выплат и др.)